articles/realtime.md

```diff
@@ -0,0 +1,330 @@
+The solution is to avoid pagefaults.
+
+Dynamic memory allocation can cause poor real-time performance.
+Calls to `malloc/new` and `free/delete` will probably result in pagefaults.
+Additionally, the heap allocates and frees memory blocks in such a way that leads to memory fragmentation, which creates poor performance for reads and writes, since the OS may have to scan for an indeterminate amount of time for a free memory block.
+
+
+### Lock memory, prefault stack:
+
+```c
+    if (mlockall(MCL_CURRENT|MCL_FUTURE) == -1) {
+            perror("mlockall failed");
+            exit(-2);
+    }
+    unsigned char dummy[MAX_SAFE_STACK];
+
+    memset(dummy, 0, MAX_SAFE_STACK);
+```
+
+[`mlockall`](http://linux.die.net/man/2/mlockall) is a Linux system call for locking the process's virtual address space into RAM, preventing the memory that will be accessed by the process from getting paged into swap space.
+
+This code snippet, when run at the beginning of a thread's lifecycle, ensures that no pagefaults occur while the thread is running.
+`mlockall` locks the stack for the thread.
+The [`memset`](http://linux.die.net/man/3/memset) call pre-loads each block of memory of the stack into the cache, so that no pagefaults will occur when the stack is accessed. (3)
+
+### Allocate dynamic memory pool
+
+```c
+  if (mlockall(MCL_CURRENT | MCL_FUTURE))
+    perror("mlockall failed:");
+
+  /* Turn off malloc trimming.*/
+  mallopt(M_TRIM_THRESHOLD, -1);
+
+  /* Turn off mmap usage. */
+  mallopt(M_MMAP_MAX, 0);
+
+  page_size = sysconf(_SC_PAGESIZE);
+  buffer = malloc(SOMESIZE);
+
+  for (i=0; i < SOMESIZE; i+=page_size)
+  {
+      buffer[i] = 0;
+  }
+  free(buffer);
+```
+
+The intro to this section stated that dynamic memory allocation is bad. However, this code snippet shows how to make dynamic memory allocation real-time safe (mostly).
+It locks the virtual address space to a fixed size, disallows returning deallocated memory to the kernel via `sbrk`, and disables `mmap`.
+This effectively locks a pool of memory in the heap into RAM, which prevents page faults due to `malloc` and `free`. (3)
+
+Pros:
+
+* Can use malloc/new, free/delete, and even STL containers
+
+Cons:
+
+* Platform/implementation dependent
+* Must accurately predict bounded memory size for the process!
+* Using STL containers is therefore dangerous (unbounded sizes)
+* In practice, only works for processes with small memory footprint
+
+### Custom static allocators for STL containers
+
+To use the standard library safely, custom allocators for STL containers must be implemented that only allocate memory on the stack.
+
+There are various comprehensive tutorials [already written](http://www.codeguru.com/cpp/article.php/c18503/C-Programming-Stack-Allocators-for-STL-Containers.htm) for this task.
+
+Pros:
+* Use existing STL code with deterministic computation
+* More modular solution
+
+Cons:
+* Complex to implement
```

Comment:
If one leverages an existing allocator, the cons end up being an additional dependency and potentially some extra complexity in setting up the allocator. The latter is allocator-specific, but typical setup parameters are pool size,  or specifying the policy of what to do if pool gets full (allocate more vs. `std::bad_alloc`). More verbose syntax (e.g. specifying allocator in STL container template params) can be alleviated with typedefs.